### DISCUSSION TITLE
Can't render a document to epub format

### DISCUSSION THREAD

**Discussion by PierreSolomon** [quarto-dev/quarto-cli#10770] [Q&A] [answered]
### Description

I have a problem when rendering some big qmd files.
I'm on windows 11, with R 4.4.1, RStudio 2024.04.2+764 and quarto 1.5.57 and I get the error next when I try to render a document to the epub format:

```
> quarto::quarto_render("./myDocument.qmd")

<--- Last few GCs --->

[15312:000001E66A012670]    46740 ms: Mark-Compact (reduce) 8160.3 (8194.3) -> 8159.0 (8194.2) MB, 650.72 / 0.00 ms  (average mu = 0.166, current mu = 0.018) allocation failure; scavenge might not succeed
[15312:000001E66A012670]    47371 ms: Mark-Compact (reduce) 8162.7 (8196.9) -> 8160.9 (8196.1) MB, 630.06 / 0.00 ms  (average mu = 0.078, current mu = 0.002) allocation failure; GC in old space requested


<--- JS stacktrace --->


#
# Fatal JavaScript out of memory: Reached heap limit
#

```

I can't share the document as there are some personal informations in it but I can say that it is only text with some basic markdown to indicate bold or italic word in the text.

I will indicate how to recreate a similar one in term of size. What is weird is that I could render this document last week and now I can't. All I did in the meantime was to update quarto from the 1.5.55 version to the 1.5.57.

My document is composed only of text. Something similar could be generated with the code next (the size is about 3 time heavier than my document but with an equal size I can render it):

```
sink("./test.qmd")
cat("---",
  "format:",
  "  epub:",
  '    title: "Test document"',
  '    author: "Lorem Ipsum"',
  "---\n\n",
  sep = "\n"
)
for (i in 1:5000) {
  cat("\n# Part", i, "\n\n", sep = "")
  cat(stringi::stri_rand_lipsum(20), sep = "\n\n")
}
sink()
quarto::quarto_render("./test.qmd")
```

Any idea of what I can do to correct this? I know that there was similar issue related to the memory allocated to deno but increasing it in powershell didn't resolve it. I tried to allocate 30gb with the command `$env:QUARTO_DENO_EXTRA_OPTIONS = "--v8-flags=--max-old-space-size=30720"`

I thank you in advance for your help.

**Comment by cderv** [comment]
This is related to this DENO error, and you need indeed to change this environment variable as describe in Troubleshoot page: https://quarto.org/docs/troubleshooting/#out-of-memory-issues

I don't how much memory your rendering requires but I would try different values, and make sure you are setting the environment properly so that `quarto render` can have the new value you want. I am giving this precision because you are using `quarto::quarto_render()` and adding a R layer. 

Try in powershell terminal directly
````
$env:QUARTO_DENO_EXTRA_OPTIONS = "--v8-flags=--max-old-space-size=30720"
quarto render test.qmd
````

**Reply by PierreSolomon** [reply]
Hello, I tried in to launch the quarto command in powershell and I get the error next:
```
> $env:QUARTO_DENO_EXTRA_OPTIONS = "--v8-flags=--max-old-space-size=30720"
> quarto render test.qmd
error: the argument '--v8-flags[=<v8-flags>...]' cannot be used multiple times

Usage: deno run [OPTIONS] [SCRIPT_ARG]...

For more information, try '--help'.
````
Look like there is a problem with the command?

What is the more strange is that like I said in my original post, just 2/3 days ago, I could render my document fine in RStudio with the quarto R package command.

**Reply by cscheid** [reply, marked as answer]
The syntax for this changed recently. V8 options have to be passed in their own special env variable now, `QUARTO_DENO_V8_OPTIONS`. You want

```
> $env:QUARTO_DENO_V8_OPTIONS = "--max-old-space-size=30720"
```

**Reply by cscheid** [reply]
(we need to document this change in https://quarto.org/docs/troubleshooting/index.html#basics), tracking on https://github.com/quarto-dev/quarto-cli/issues/10772

**Reply by PierreSolomon** [reply]
This work with this option. Is it possible to set it in the RStudio instance? Maybe as a parameter in the quarto_render function? It would be useful for the workflow I use.

**Reply by cderv** [reply]
> is it possible to set it in the RStudio instance?

You mean on a server instance ? If so, you can look at how to configure at server level, or product level. 

> Maybe as a parameter in the quarto_render function?

You can set the environment in your R environment and it should then be set correctly for when Quarto is called from R.

**Reply by mcanouil** [reply]
I would suggest to use `_environment` in your project as described in https://quarto.org/docs/projects/environment.html

**Reply by cderv** [reply]
@mcanouil This would be indeed quite a good configuration if this works ! Did you try it confirm ? 

I did not suggest it because `_environment` is a Quarto feature, meant to have some environment variable available for `quarto render` and other commands. But for this file to be loaded and read, quarto must be running, which means Deno must be running. 

`QUARTO_DENO_V8_OPTIONS` is passed to the command line calling `deno` which calls `quarto.js` script. So way ahead of `_environment` feature to be activated. It's why I would be surprised if this work. But it would be a good surprise, and a wrong guessed behavior from me !

**Reply by mcanouil** [reply]
True, I am not sure it works since I don't know how to check it is taken into account.

I can't repro an issue but it's taking way too much time to render.

**Reply by cderv** [reply]
Ok. As you suggested that, I thought it was validated solution. So my best guess is that it won't work. `QUARTO_DENO_V8_OPTIONS` needs to be set before `quarto` is called. 

@PierreSolomon will confirm probably testing the different solution.

**Reply by mcanouil** [reply]
Sorry, I wan't clear. It was really a suggestion worth trying as I am not sure when `_environment` is read.

**Comment by PierreSolomon** [comment]
Hello, `_environment`doesn't seem to work so I ended to add the `Sys.setenv(QUARTO_DENO_V8_OPTIONS = "--max-old-space-size=10000")` function in a `.Rprofile` file.

**Reply by mcanouil** [reply]
This would only work if you run Quarto through R via the R package wrapper.
For instance, if you do `quarto render`, it won't have the environment variable.
